index.js

```diff
@@ -4,15 +4,155 @@
+  var $zero = 'fantasy-land/zero';
+  var $$type = '@@type';
+
+  var ordinal = ['first', 'second', 'third', 'fourth', 'fifth'];
+
+  function isFunction(f){
+    return typeof f === 'function';
+  }
+
+  function isBinary(f){
+    return f.length >= 2;
+  }
+
+  function isApplicativeRepr(Repr){
+    try{
+      return Z.Applicative.test(Z.of(Repr));
+    }catch(_){
+      return false;
+    }
+  }
+
+  function invalidArgument(it, at, expected, actual){
+    throw new TypeError(
+      it
+      + ' expects its '
+      + ordinal[at]
+      + ' argument to '
+      + expected
+      + '\n  Actual: '
+      + Z.toString(actual)
+    );
+  }
+
+  function invalidContext(it, actual, an){
+    throw new TypeError(
+      it + ' was invoked outside the context of a ' + an + '. \n  Called on: ' + Z.toString(actual)
+    );
+  }
+
+  function last(xs){
+    return xs[xs.length - 1];
+  }
+
+  function getReprType(Repr){
+    return Repr[$$type] || Repr.name || 'Anonymous';
+  }
+
+  function concurrentPrepender(x, i, xs){
+    return i === xs.length - 1 ? ('Concurrent' + x) : x;
+  }
+
+  function computeType(type){
+    return type.split('/').map(concurrentPrepender).join('/');
+  }
+
+  //concurrify :: Applicative m
+  //           => (TypeRep m, () -> m a, (m a, m a) -> m a, (m a, m (a -> b)) -> m b)
+  //           -> Concurrently m
+  return function concurrify(Repr, zero, race, ap){
+
+    var INNERTYPE = getReprType(Repr);
+    var OUTERTYPE = computeType(INNERTYPE);
+
+    function Concurrently(sequential){
+      this.sequential = sequential;
+    }
+
+    function isInner(x){
+      return x instanceof Repr
+      || (Boolean(x) && x.constructor === Repr)
+      || type(x) === Repr[$$type];
+    }
+
+    function isOuter(x){
+      return x instanceof Concurrently
+      || (Boolean(x) && x.constructor === Concurrently)
+      || type(x) === OUTERTYPE;
+    }
+
+    function construct(x){
+      if(!isInner(x)) invalidArgument(OUTERTYPE, 0, 'be of type "' + INNERTYPE + '"', type(x));
+      return new Concurrently(x);
+    }
+
+    if(!isApplicativeRepr(Repr)) invalidArgument('concurrify', 0, 'represent an Applicative', Repr);
+
+    if(!isFunction(zero)) invalidArgument('concurrify', 1, 'be a function', zero);
+    var _zero = zero();
+    if(!isInner(_zero)) invalidArgument('concurrify', 1, 'to return ' + INNERTYPE, type(_zero));
+
+    if(!isFunction(race)) invalidArgument('concurrify', 2, 'be a function', race);
+    if(!isBinary(race)) invalidArgument('concurrify', 2, 'be binary', race);
+    var _race = race(_zero, _zero);
+    if(!isInner(_race)) invalidArgument('concurrify', 2, 'to return ' + INNERTYPE, type(_race));
+
+    if(!isFunction(ap)) invalidArgument('concurrify', 3, 'be a function', ap);
+    if(!isBinary(ap)) invalidArgument('concurrify', 3, 'be binary', ap);
+    var _ap = ap(_zero, Repr[$of](function(){}));
+    if(!isInner(_ap)) invalidArgument('concurrify', 2, 'to return ' + INNERTYPE, type(_ap));
```

Comment:
We can "have a cake and eat it too", instead of checking it on creation, do it lazily in appropriate method as you commented, except have some flag to fix performance:
```js
let isZeroChecked = false
construct[$zero] = function Concurrently$zero() {
  var _zero = zero()
  if (isZeroChecked == false) {
    // it's is first invocation of `zero` so we check if it's inner
    if (isInner(_zero)) {
      // and if it is we update `isZeroChecked` so we don't check it again
      isZeroChecked = true
    } else {
      // else boom!
      throw new TypeError()
    }
  }
  return new Concurrently(_zero)
}
```